File: tutorial/05-arrays/problems/12-smallest-positive-missing/soft.java
class Soft05p12 {
     public void sort (int [] arr,int start,int end){
        if(start<end){
            int mid= start+(end-start)/2;
            this.sort(arr,start,mid);
            this.sort(arr,mid+1,end);
            this.merge(arr, start, mid, end);
        }
    }
    public void merge(int []arr,int start,int mid,int end){
        int i=0,j=0,current=start;
        int temp1Length=mid-start+1;
        int temp2Length=end-mid;
        int []temp1=new int[temp1Length];
        int[]temp2=new int[temp2Length];
        for(int m=0;m<temp1Length;m++){
            temp1[m]=arr[start+m];
        }
        for(int m=0;m<temp2Length;m++){
            temp2[m]=arr[mid+1+m];
        }
        while (i < temp1Length && j < temp2Length){
            if(temp1[i] < temp2[j])
            {
                arr[current]=temp1[i];
                i++;
            }else
             {
                 arr[current]=temp2[j];
                 j++;
             }
             current++;
        }
        while(i<temp1Length){
            arr[current]=temp1[i];
            i++;
            current++;
        }
        while(j<temp2Length){
            arr[current]=temp2[j];
        j++;
        current++;
        }
    }
    public int missingNumber(int[] arr) {
        // code here
        int length=arr.length;
        int smallPositive=1;
        this.sort(arr,0,length-1);
        for (int i=0;i<length;i++){
            if(arr[i]==smallPositive)
            {
                smallPositive++;
            }
        }
        return smallPositive ;
    }
    public static void main(String[] args) {
        Soft05p12 soft=new Soft05p12();
        int []arr= new int[]{2, -3, 4, 1, 1, 7};
        System.out.println(soft.missingNumber(arr));
    }
}

File: tutorial/05-arrays/problems/12-smallest-positive-missing/soft.py
class Solution:
    def sort(self,arr,leftIndex,rightIndex):
        if leftIndex<rightIndex:
            mid= leftIndex+((rightIndex-leftIndex)//2)
            self.sort(arr,leftIndex,mid)
            self.sort(arr,mid+1,rightIndex)
            self.merge(arr,leftIndex,mid,rightIndex)
    def merge(self,arr,leftIndex,mid,rightIndex):
        i=j=0
        current=leftIndex
        temp1Length=mid-leftIndex+1
        temp2Length=rightIndex-mid
        temp1=arr[leftIndex:leftIndex+temp1Length]
        temp2=arr[mid+1:mid+1+temp2Length]
        while i < temp1Length and j < temp2Length:
            if temp1[i] < temp2[j]:
                arr[current]=temp1[i]
                i+=1
            else:
                arr[current]=temp2[j]
                j+=1
            current+=1
        while i < temp1Length:
            arr[current]=temp1[i]
            i+=1
            current+=1
        while j<temp2Length:
            arr[current]=temp2[j]
            j+=1
            current+=1
    def missingNumber(self, arr):
        length=len(arr)
        print('length:',length)
        self.sort(arr,0,length-1)
        print(arr)
        smallPositive=1
        for i in range(length):
            if smallPositive==arr[i]:
                smallPositive+=1
        return smallPositive
soft = Solution()
arr=[2, -3, 4, 1, 1, 7]
print(soft.missingNumber(arr))